File: src/main/go/leetcode/lists/hot100/LeetCode208.go
package hot100

type Trie struct {
	children [26]*Trie
	isEnd    bool
}

func ConstructorTrie() Trie {
	return Trie{
		children: [26]*Trie{},
		isEnd:    false,
	}
}

func (this *Trie) Insert(word string) {
	cur := this
	for i := range word {
		ch := word[i]
		if cur.children[ch-'a'] == nil {
			cur.children[ch-'a'] = &Trie{}
		}
		cur = cur.children[ch-'a']
	}
	cur.isEnd = true
}

func (this *Trie) Search(word string) bool {
	if len(word) == 0 {
		return true
	}
	curr := this
	for i := 0; i < len(word); i++ {
		ch := word[i]
		if curr.children[ch-'a'] == nil {
			return false
		}
		curr = curr.children[ch-'a']
	}
	return curr.isEnd
}

func (this *Trie) StartsWith(prefix string) bool {
	if len(prefix) == 0 {
		return true
	}
	curr := this
	for i := 0; i < len(prefix); i++ {
		ch := prefix[i]
		if curr.children[ch-'a'] == nil {
			return false
		}
		curr = curr.children[ch-'a']
	}
	return true
}

File: java/src/leetcode/LeetCode208.java
package leetcode;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/17 16:30
 */
public class LeetCode208 {

    class Trie {

        TrieNode root;

        public Trie() {
            root = new TrieNode();
        }

        public void insert(String word) {
            char[] chars = word.toCharArray();
            TrieNode node = root;
            for (int i = 0; i < chars.length; i++) {
                if (node.children[chars[i] - 'a'] == null) {
                    node.children[chars[i] - 'a'] = new TrieNode();
                }
                node = node.children[chars[i] - 'a'];
            }
            node.isEnd = true;
        }

        public boolean search(String word) {
            char[] chars = word.toCharArray();
            TrieNode node = root;
            for (int i = 0; i < chars.length; i++) {
                if (node.children[chars[i] - 'a'] == null) {
                    return false;
                }
                node = node.children[chars[i] - 'a'];
            }
            return node.isEnd;
        }

        public boolean startsWith(String prefix) {
            char[] chars = prefix.toCharArray();
            TrieNode node = root;
            for (int i = 0; i < chars.length; i++) {
                if (node.children[chars[i] - 'a'] == null) {
                    return false;
                }
                node = node.children[chars[i] - 'a'];
            }
            return true;
        }

        static class TrieNode {
            public TrieNode[] children = new TrieNode[26];
            boolean isEnd = false;
        }
    }

}
